### PR TITLE
Fix mm accuracy in ROCm for some inputs

### PR DESCRIPTION
**Pull request by lcskrishna**
This PR fixes the accuracy issues for hipblasLT for mm case on ROCm. 
This PR is a follow up to the integration PR https://github.com/pytorch/pytorch/pull/114329 and https://github.com/pytorch/pytorch/pull/114890

The accuracy issue arises for mm usecase for ROCm where hipblasLT is enabled, and a bias has been passed which is not required. This PR addresses that issue. 
Added a unit-test case for this issue (bias=None) case. 

cc @jeffdaily @sunway513 @jithunnair-amd @pruthvistony @ROCmSupport @dllehr-amd @jataylo @hongxiayang

### CODE REVIEW

File: aten/src/ATen/native/cuda/Blas.cpp [hunk 1]
```diff
@@ -339,7 +339,11 @@ Tensor& addmm_out_cuda_impl(Tensor& result, const Tensor& self, const Tensor& ma
               args.lda,
               args.matb->data_ptr<scalar_t>(),
               args.ldb,
+#if defined(USE_ROCM)
+              (&result != &self) ? self.const_data_ptr<scalar_t>() : nullptr,
```
Comment (cpuhrsch) [inline aten/src/ATen/native/cuda/Blas.cpp:343]: Is there a way of writing this more lines of code, but more readable? This isn't exactly clear on a first read and this code is important enough that I worry that sort of complexity will make it confusing.

File: aten/src/ATen/cuda/CUDABlas.cpp [hunk 2]
```diff
@@ -745,8 +745,11 @@ void gemm_and_bias(
     epilogue = CUBLASLT_EPILOGUE_GELU_BIAS;
 #endif
   }
-  computeDesc.setAttribute(CUBLASLT_MATMUL_DESC_EPILOGUE, epilogue);
-  computeDesc.setAttribute(CUBLASLT_MATMUL_DESC_BIAS_POINTER, bias);
+
+  if (bias) {
```
Comment (cpuhrsch) [inline aten/src/ATen/cuda/CUDABlas.cpp:749]: I see, you're trying to use nullptr to indicate optionality.

Comment (cpuhrsch) [reply aten/src/ATen/cuda/CUDABlas.cpp:749]: Just for clarity I'd prefer an explicit comparison with nullptr and a comment.

Having said that, we usually use a different pattern (the optional utility) to indicate optionality.

I'm not sure if we want to insist on that here. cc @albanD @malfet

Comment (albanD) [reply aten/src/ATen/cuda/CUDABlas.cpp:749]: In these very low level kernels where everything is raw pointers, we don't usually use Optional anymore.

**Review by albanD — COMMENTED** [review]
A more detailed PR description and a test case in OpInfo that triggers this failure would be great.

File: aten/src/ATen/cuda/CUDABlas.cpp (hunk 2 above)
Comment (lcskrishna) [reply aten/src/ATen/cuda/CUDABlas.cpp:749]: @cpuhrsch  @albanD  Let me check against nullptr to make it more readable.

File: aten/src/ATen/native/cuda/Blas.cpp (hunk 1 above)
Comment (lcskrishna) [reply aten/src/ATen/native/cuda/Blas.cpp]: Done.

**Review by jeffdaily — APPROVED** [review]

File: test/test_linalg.py [hunk 3]
```diff
@@ -4682,6 +4683,33 @@ def test_corner_cases_of_cublasltmatmul(self, device, dtype):
         m2 = torch.randn(16, 131071, device=device).to(dtype)
         torch.nn.functional.linear(m1, m2, M)
 
+    @skipCUDAIfNotRocm
+    @dtypes(*floating_types_and(torch.bfloat16, torch.half))
+    def test_hipblaslt_corner_cases_rocm(self, device, dtype):
+        # enable hipblaslt path via env variable.
+        import os
+        os.environ["DISABLE_ADDMM_HIP_LT"] = "0"
+        if (dtype != torch.double):
```
Comment (malfet) [inline test/test_linalg.py:4692]: IMO if the intent to skip the testing for double types, it would be better to say something like:
```python
   if type == torch.double:
     raise unittest.skipTest("Not yet supported for doubles")

**Review by malfet — APPROVED** [review]
LGTM, but see comments about some minor issues (but as tests are limited to ROCm platform I'm not requesting changes)

File: test/test_linalg.py [hunk 4]
```diff
@@ -4682,6 +4683,33 @@ def test_corner_cases_of_cublasltmatmul(self, device, dtype):
...
+            out1 = torch.nn.functional.linear(m1, m2, M)
+            M_cpu = M.to('cpu')
+            m1_cpu = m1.to('cpu')
+            m2_cpu = m2.to('cpu')
+            out1_cpu = torch.nn.functional.linear(m1_cpu, m2_cpu, M_cpu)
+            self.assertTrue(torch.allclose(out1_cpu, out1.cpu(), rtol=1e-2, atol=1e-2))
+
+            # common case without bias
+            m1 = torch.randn(2048, 2400, device=device).to(dtype)
+            m2 = torch.randn(128, 2400, device=device).to(dtype)
```
Comment (malfet) [inline test/test_linalg.py:4706]:
```suggestion
            m1 = torch.randn(2048, 2400, device=device, dtype=dtype)
            m2 = torch.randn(128, 2400, device=device, dtype=dtype)
```

File: test/test_linalg.py (hunk 3 above)
Comment (lcskrishna) [reply test/test_linalg.py]: Thanks, this makes sense. Let me update it.

**Comment by xw285cornell** [comment]
Any chance we can change the hipblaslt behavior to avoid cuda/hip divergence?

**Comment by jeffdaily** [comment]
> Any chance we can change the hipblaslt behavior to avoid cuda/hip divergence?

Which part of it?  The bias issue or the double type not being supported or both?

File: aten/src/ATen/native/cuda/Blas.cpp [hunk 5]
```diff
@@ -339,7 +339,13 @@ Tensor& addmm_out_cuda_impl(Tensor& result, const Tensor& self, const Tensor& ma
               args.lda,
               args.matb->data_ptr<scalar_t>(),
               args.ldb,
+#if defined(USE_ROCM)
+              // This condition is needed for mm case on ROCm for hipblasLt path.
+              // Passing the bias ptr as null to avoid accuracy issues for mm case.
+              (&result != &self) ? self.const_data_ptr<scalar_t>() : nullptr,
```
Comment (xw285cornell) [inline aten/src/ATen/native/cuda/Blas.cpp:345]: @jeffdaily I meant here that uses bias==nullptr to avoid setting the attributes in computeDesc iiuc. I wonder why setting those epilog attr will end up with wrong results.

**Comment by jeffdaily** [comment]
pytorchmergebot got confused.  I had to remove and re-add ciflow labels.  Hopefully all missing ciflows are available to mergebot now.
